### PR TITLE
Catch missing classlike exceptions during scanning

### LINKED ISSUE #10706: 5.22.0 regression: Could not get class storage for psl\range\upperboundrangeinterface
It was fine in 5.21.1

```
Target PHP version: 8.3 (inferred from current PHP version).
Scanning files...
Uncaught InvalidArgumentException: Could not get class storage for psl\range\upperboundrangeinterface in /app-appname/vendor/vimeo/psalm/src/Psalm/Internal/Provider/ClassLikeStorageProvider.php:45
Stack trace:
#0 /app-appname/vendor/vimeo/psalm/src/Psalm/Internal/Codebase/ClassLikes.php(726): Psalm\Internal\Provider\ClassLikeStorageProvider->get()
#1 /app-appname/vendor/vimeo/psalm/src/Psalm/Internal/Codebase/ClassLikes.php(716): Psalm\Internal\Codebase\ClassLikes->getParentInterfaces()
#2 /app-appname/vendor/vimeo/psalm/src/Psalm/Codebase.php(763): Psalm\Internal\Codebase\ClassLikes->interfaceExtends()
#3 /app-appname/vendor/vimeo/psalm/src/Psalm/Internal/Type/Comparator/ObjectComparator.php(357): Psalm\Codebase->interfaceExtends()
#4 /app-appname/vendor/vimeo/psalm/src/Psalm/Internal/Type/Comparator/ObjectComparator.php(117): Psalm\Internal\Type\Comparator\ObjectComparator::isIntersectionShallowlyContainedBy()
#5 /app-appname/vendor/vimeo/psalm/src/Psalm/Internal/Type/Comparator/AtomicTypeComparator.php(350): Psalm\Internal\Type\Comparator\ObjectComparator::isShallowlyContainedBy()
#6 /app-appname/vendor/vimeo/psalm/src/Psalm/Type.php(891): Psalm\Internal\Type\Comparator\AtomicTypeComparator::isContainedBy()
#7 /app-appname/vendor/vimeo/psalm/src/Psalm/Type.php(768): Psalm\Type::intersectAtomicTypes()
#8 /app-appname/vendor/vimeo/psalm/src/Psalm/Internal/PhpVisitor/Reflector/TypeHintResolver.php(112): Psalm\Type::intersectUnionTypes()
#9 /app-appname/vendor/vimeo/psalm/src/Psalm/Internal/PhpVisitor/Reflector/FunctionLikeNodeScanner.php(405): Psalm\Internal\PhpVisitor\Reflector\TypeHintResolver::resolve()
#10 /app-appname/vendor/vimeo/psalm/src/Psalm/Internal/PhpVisitor/ReflectorVisitor.php(207): Psalm\Internal\PhpVisitor\Reflector\FunctionLikeNodeScanner->start()
#11 /app-appname/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php(200): Psalm\Internal\PhpVisitor\ReflectorVisitor->enterNode()
#12 /app-appname/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php(114): PhpParser\NodeTraverser->traverseArray()
#13 /app-appname/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php(223): PhpParser\NodeTraverser->traverseNode()
#14 /app-appname/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php(114): PhpParser\NodeTraverser->traverseArray()
#15 /app-appname/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php(223): PhpParser\NodeTraverser->traverseNode()
#16 /app-appname/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php(91): PhpParser\NodeTraverser->traverseArray()
#17 /app-appname/vendor/vimeo/psalm/src/Psalm/Internal/Scanner/FileScanner.php(79): PhpParser\NodeTraverser->traverse()
#18 /app-appname/vendor/vimeo/psalm/src/Psalm/Internal/Codebase/Scanner.php(554): Psalm\Internal\Scanner\FileScanner->scan()
#19 /app-appname/vendor/vimeo/psalm/src/Psalm/Internal/Codebase/Scanner.php(782): Psalm\Internal\Codebase\Scanner->scanFile()
#20 /app-appname/vendor/vimeo/psalm/src/Psalm/Internal/Codebase/Scanner.php(428): Psalm\Internal\Codebase\Scanner->scanAPath()
#21 /app-appname/vendor/vimeo/psalm/src/Psalm/Internal/Codebase/Scanner.php(280): Psalm\Internal\Codebase\Scanner->scanFilePaths()
#22 /app-appname/vendor/vimeo/psalm/src/Psalm/Config.php(2576): Psalm\Internal\Codebase\Scanner->scanFiles()
#23 /app-appname/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/ProjectAnalyzer.php(372): Psalm\Config->visitComposerAutoloadFiles()
#24 /app-appname/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/ProjectAnalyzer.php(502): Psalm\Internal\Analyzer\ProjectAnalyzer->visitAutoloadFiles()
#25 /app-appname/vendor/vimeo/psalm/src/Psalm/Internal/Cli/Psalm.php(374): Psalm\Internal\Analyzer\ProjectAnalyzer->check()
#26 /app-appname/vendor/vimeo/psalm/psalm(9): Psalm\Internal\Cli\Psalm::run()
#27 /app-appname/vendor/bin/psalm(119): include('...')
#28 {main}
(Psalm 5.22.0@fe2c67ec89f358940f90db05efd2d663388b45a6 crashed due to an uncaught Throwable)

```

### PR DESCRIPTION
**Pull request by weirdan**
This may resolve #10706, although I'm not exactly convinced it's the
best way.

Fixes: #7520 
Fixes: #10350
Fixes: #10152

### CODE REVIEW

**Comment by ohader** [comment]
@weirdan Thanks for picking this up. Your PR triggered me to dig deeper into this scenario. I've created an alternative PR at #10724 which introduces "rescanning"... 

Basically any invocation of [`AtomicTypeComparator::isContainedBy()`](https://github.com/vimeo/psalm/blob/c9493c54cdffe87b5e19baeb2ce9aade7093059d/src/Psalm/Internal/Type/Comparator/AtomicTypeComparator.php#L56) could lead to the described class-like failure.

File: src/Psalm/Type.php [hunk 1]
```diff
@@ -888,33 +888,37 @@ private static function intersectAtomicTypes(
...
-                && !self::hasIntersection($type_2_atomic)
-            ) {
-                return $intersection_atomic;
+                if ($intersection_atomic
+                    && !self::hasIntersection($type_1_atomic)
+                    && !self::hasIntersection($type_2_atomic)
+                ) {
+                    return $intersection_atomic;
+                }
+            } catch (InvalidArgumentException $e) {
```
Comment (ohader) [inline src/Psalm/Type.php:920]: Muting this exception would not address the underlying issue: The class-like storage items (e.g. for `\JsonSerializable` or `\Stringable` and any other external interface) not being available (yet) at that time during the scanning process..

Comment (weirdan) [reply src/Psalm/Type.php:920]: The question is, do we get broken data (e.g. `Countable&Iterator` turning into `Countable`), or do we just not get as tight an intersection as possible (e.g. `MySessionHandler&SessionHandlerInterface` not simplified to `MySessionHandler`). If it's the latter, it's probably not the end of the world and would be handled during the analysis phase at the cost of some wasted CPU cycles.

**Review by ohader — APPROVED** [review]
_before:_

Screenshot 2024-02-21 at 13 01 49

_after:_

Screenshot 2024-02-21 at 13 16 15

The screenshots above visualize the before/after scenario of our tester for TYPO3-specific 3rd party packages (enriched by custom stubs and PsalmPHP hooks/events).

---

Let's go with this change, since it actually fixes the main problem on those intersections.

I was too optimistic with my alternative approaches and failed with other side effects in process forks (we'd probably would need something like `Promise(ClassLikeStorage)` in async PHP-environments).
In addition some tests failed when getting rid of that exception - and probably other code components relied on that as well. 

Thus, my conclusion is, that it would be way too risky to fiddle around with these parts and only adjust the those parts we can isolate better (like the `Psalm\Type` class of this PR).

**Comment by ohader** [comment]
@weirdan Can you please cherry-pick [my test cases](https://github.com/ohader/psalm/commit/85a82b83f16034b43035cb0e63f7ebfa37529f92) to the top of your branch?

```
git fetch https://github.com/ohader/psalm.git 10706-tests && git cherry-pick FETCH_HEAD
```

**Pull request merged**